### PR TITLE
[NX-51] Add support for create new payment endpoint

### PR DESCRIPTION
**Pull request by aplegatt**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1564,20 +1592,11 @@ await client.checkout.orderNext({ orderToken })
 // Step three - pick a payment method
 const payment = (await client.checkout.paymentMethods({ orderToken })).success()
 
-await client.checkout.orderUpdate({ orderToken }, {
+await client.checkout.addPayment({ orderToken }, {
   order: {
     payments_attributes: [{
       payment_method_id: payment.data[0].id
     }]
```
Comment (damianlegawiec) [inline README.md:1599]: @aplegatt this is wrong payload

File: README.md [hunk 2]
```diff
@@ -1256,6 +1256,63 @@ const response = await client.checkout.shippingMethods(
...
+
+**Success response schema:** [Success schema](#success-schema)
+
+**Failure response schema:** [Error schema](#error-schema)
+
+**Example:**
+
+```ts
+// Logged in user
+const response = await client.checkout.addPayment(
```
Comment (damianlegawiec) [inline README.md:1292]: @aplegatt we need an example for existing credit card, as it's described in Spree API docs

File: README.md [hunk 3]
```diff
@@ -1256,6 +1256,63 @@ const response = await client.checkout.shippingMethods(
...
+  { bearerToken },
+  {
+    payment_method_id: 2,
+    source_attributes: {
+      gateway_payment_profile_id: "card_1JqvNB2eZvKYlo2C5OlqLV7S",
+      name: "John Snow"
```
Comment (damianlegawiec) [inline README.md:1298]: @aplegatt this need to include cc_type/last_digits/month/year as it's described in the API docs - https://api.spreecommerce.org/docs/api-v2/b3A6MjYyODA2NTY-create-new-payment


Screenshot 2021-11-02 at 17 23 36

**Review by damianlegawiec — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -1256,6 +1256,93 @@ const response = await client.checkout.shippingMethods(
...
+)
+
+// Use existing credit card
+const response = await client.checkout.addPayment(
+  { orderToken },
+  {
+    payment_method_id: "1",
+    source_id: "1"
+  }
+)
```
Comment (damianlegawiec) [inline README.md:1343]: @aplegatt guest users can only add a new card, they cannot use existing ones, security reasons :)

**Review by damianlegawiec — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -1564,20 +1622,15 @@ await client.checkout.orderNext({ orderToken })
...
-    }
+await client.checkout.addPayment({ orderToken }, {
+  payment_method_id: payment.data[0].id,
+  source_attributes: {
+    gateway_payment_profile_id: payment.data[0].source.gateway_payment_profile_id,
+    cc_type: payment.data[0].source.cc_type,
+    last_digits: payment.data[0].source.last_digits,
+    month: payment.data[0].source.month,
+    year: payment.data[0].source.year,
+    name: payment.data[0].source.name
```
Comment (damianlegawiec) [inline README.md:1633]: @aplegatt this data is not coming from payment method - this is a payment, a completely different entity, this needs to be set by hand

**Review by damianlegawiec — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (aplegatt) [reply README.md:1633]: @damianlegawiec How should I present it in the readme file then?

Comment (damianlegawiec) [reply README.md:1633]: @aplegatt use the example data - https://github.com/spree/spree-storefront-api-v2-js-sdk/pull/273/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R1326-R1331

**Review by damianlegawiec — CHANGES_REQUESTED** [review]
@aplegatt 

1. This endpoint is missing from the table of contents in the beginning of the readme
2. Please remove any references to passing `payments_attributes` and `payment_source` in the checkout.orderUpdate endpoint

**Review by damianlegawiec — CHANGES_REQUESTED** [review]
@aplegatt `one step` / `three steps` links should be removed
Screenshot 2021-11-03 at 13 05 46

**Review by tniezg — COMMENTED** [review]
`tests/e2e/spec.ts` is incorrectly calling `client.checkout.orderUpdate`. The test needs to be updated to reflect the new `orderUpdate` signature:

image

File: README.md [hunk 6]
```diff
@@ -1256,6 +1241,84 @@ const response = await client.checkout.shippingMethods(
...
+
+**Required token:** [Bearer token](#bearer-token) or [Order token](#order-token)
+
+**Parameters schema:**
+
+```ts
+payment_method_id: string
+source_id?: string
+amount?: number
+source_attributes: {
```
Comment (tniezg) [inline README.md:1256]:
```suggestion
source_attributes?: {
```

File: src/interfaces/endpoints/CheckoutClass.ts [hunk 7]
```diff
@@ -8,14 +8,30 @@ export interface AddStoreCredit extends IQuery {
...
-    payments_attributes?: IPayment[]
+    payments_attributes?: AddFullPayment
     shipments_attributes?: IShipment[]
   }
-  payment_source?: IPaymentSource
+}
+
+export interface AddFullPayment extends OrderUpdate {
+  payment_method_id: string
+  source_attributes: {
```
Comment (tniezg) [inline src/interfaces/endpoints/CheckoutClass.ts:24]:
```suggestion
  source_attributes?: {
```

File: src/interfaces/endpoints/CheckoutClass.ts [hunk 8]
```diff
@@ -8,14 +8,30 @@ export interface AddStoreCredit extends IQuery {
...
     bill_address_attributes?: IAddress
     ship_address_attributes?: IAddress
-    payments_attributes?: IPayment[]
+    payments_attributes?: AddFullPayment
     shipments_attributes?: IShipment[]
   }
-  payment_source?: IPaymentSource
+}
+
+export interface AddFullPayment extends OrderUpdate {
```
Comment (tniezg) [inline src/interfaces/endpoints/CheckoutClass.ts:22]:
```suggestion
export interface AddFullPayment {
```

File: src/interfaces/endpoints/CheckoutClass.ts [hunk 9]
```diff
@@ -8,14 +8,30 @@ export interface AddStoreCredit extends IQuery {
...
+    gateway_payment_profile_id: string
+    cc_type?: string
+    last_digits?: string
+    month?: string
+    year?: string
+    name: string
+  }
+}
+
+export interface AddPayment extends AddFullPayment {
```
Comment (tniezg) [inline src/interfaces/endpoints/CheckoutClass.ts:34]:
```suggestion
export interface AddPayment extends AddFullPayment, IQuery {
```

File: src/interfaces/endpoints/CheckoutClass.ts [hunk 10]
```diff
@@ -8,14 +8,30 @@ export interface AddStoreCredit extends IQuery {
...
 
-export interface NestedAttributes extends IQuery {
+export interface OrderUpdate extends IQuery {
   order?: {
     email?: string
     special_instructions?: string
     bill_address_attributes?: IAddress
     ship_address_attributes?: IAddress
-    payments_attributes?: IPayment[]
+    payments_attributes?: AddFullPayment
```
Comment (tniezg) [inline src/interfaces/endpoints/CheckoutClass.ts:17]:
```suggestion
    payments_attributes?: AddFullPayment[]
```

**Review by tniezg — COMMENTED** [review]
Please remove the `IPaymentSource` type and `IPayment`. These are no longer used.

**Comment by damianlegawiec** [comment]
> Please remove the `IPaymentSource` type and `IPayment` types. These are no longer used.

maybe rather than removing let's add a deprecation notice

**Comment by tniezg** [comment]
> > Please remove the `IPaymentSource` type and `IPayment` types. These are no longer used.
> 
> maybe rather than removing let's add a deprecation notice

@damianlegawiec Depreciating the types while they're no longer used anywhere is not going to be very helpful. If you really want to, we can allow the old types along with new types when making the request to `orderUpdate`. We tend not to depreciate, because switching between most SDK versions is easy and adding or changing endpoints is easy too.

**Review by tniezg — APPROVED** [review]

**Pull request merged**